hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java

```diff
@@ -183,105 +270,166 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
-              // Write the cell to a del file
-              delFileWriter.append(c);
-              deleteMarkersCount++;
+              // Not a MOB reference cell
+              int size = c.getValueLength();
+              if (size > mobSizeThreshold) {
+                // This MOB cell comes from a regular store file
+                // therefore we store it in original mob output
+                mobFileWriter.append(c);
+                writer
+                    .append(MobUtils.createMobRefCell(c, fileName, this.mobStore.getRefCellTags()));
+                mobCells++;
+              } else {
+                writer.append(c);
+              }
             }
-          } else if (mobFileWriter == null || c.getTypeByte() != KeyValue.Type.Put.getCode()) {
-            // If the mob file writer is null or the kv type is not put, directly write the cell
+          } else if (c.getTypeByte() != KeyValue.Type.Put.getCode()) {
+            // Not a major compaction or major with MOB disabled
+            // If the kv type is not put, directly write the cell
             // to the store file.
             writer.append(c);
           } else if (MobUtils.isMobReferenceCell(c)) {
+            // Not a major MOB compaction, Put MOB reference
             if (MobUtils.hasValidMobRefCellValue(c)) {
               int size = MobUtils.getMobValueLength(c);
               if (size > mobSizeThreshold) {
                 // If the value size is larger than the threshold, it's regarded as a mob. Since
                 // its value is already in the mob file, directly write this cell to the store file
                 writer.append(c);
+                // Add MOB reference to a set
+                mobRefSet.get().add(MobUtils.getMobFileName(c));
               } else {
                 // If the value is not larger than the threshold, it's not regarded a mob. Retrieve
-                // the mob cell from the mob file, and write it back to the store file. Must
-                // close the mob scanner once the life cycle finished.
-                try (MobCell mobCell = mobStore.resolve(c, false)) {
-                  if (mobCell.getCell().getValueLength() != 0) {
-                    // put the mob data back to the store file
-                    PrivateCellUtil.setSequenceId(mobCell.getCell(), c.getSequenceId());
-                    writer.append(mobCell.getCell());
-                    cellsCountCompactedFromMob++;
-                    cellsSizeCompactedFromMob += mobCell.getCell().getValueLength();
-                  } else {
-                    // If the value of a file is empty, there might be issues when retrieving,
-                    // directly write the cell to the store file, and leave it to be handled by the
-                    // next compaction.
-                    writer.append(c);
-                  }
+                // the mob cell from the mob file, and write it back to the store file.
```

Comment:
why are we doing  this during a non-mob compaction? Shouldn't we just treat all the mob cells as pass through?

we could simplify a bunch of this logic if for non-mob compactions we didn't worry about moving things in or out of the mob area.

Reply:
1. MOB threshold could be changed and we will need to move MOB cell to/from MOB file
2. Store file could have value which is larger than MOB threshold and we will need to take care of that as well.

3. It works.